trio/_subprocess.py

```diff
@@ -592,44 +603,49 @@ async def my_deliver_cancel(process):
     stdout_chunks = []
     stderr_chunks = []
 
-    async with await open_process(command, **options) as proc:
-
-        async def feed_input():
-            async with proc.stdin:
-                try:
-                    await proc.stdin.send_all(input)
-                except trio.BrokenResourceError:
-                    pass
+    async def feed_input(stream):
+        async with stream:
+            try:
+                await stream.send_all(input)
+            except trio.BrokenResourceError:
+                pass
 
-        async def read_output(stream, chunks):
-            async with stream:
-                async for chunk in stream:
-                    chunks.append(chunk)
+    async def read_output(stream, chunks):
+        async with stream:
+            async for chunk in stream:
+                chunks.append(chunk)
 
-        async with trio.open_nursery() as nursery:
-            if proc.stdin is not None:
-                nursery.start_soon(feed_input)
-            if proc.stdout is not None:
+    async with trio.open_nursery() as nursery:
+        proc = await open_process(command, **options)
+        try:
+            if input is not None:
+                nursery.start_soon(feed_input, proc.stdin)
+                proc.stdin = None
+                proc.stdio = None
+            if capture_stdout:
                 nursery.start_soon(read_output, proc.stdout, stdout_chunks)
-            if proc.stderr is not None:
+                proc.stdout = None
+                proc.stdio = None
+            if capture_stderr:
                 nursery.start_soon(read_output, proc.stderr, stderr_chunks)
-            try:
+                proc.stderr = None
+            task_status.started(proc)
+            await proc.wait()
+        except BaseException:
+            with trio.CancelScope(shield=True):
```

Comment:
Yeah this is great; love how the spawning maps 1-to-1 with a task now.

This is exactly what we [ended up with in `tractor`](https://github.com/goodboy/tractor/blob/master/tractor/_spawn.py#L232) since we also needed a shielded `Process.wait()` before the call to `Process.__aexit__()` (internal IPC protocol takes care of inter-process cancellation) and the task mapping pattern was the only way to reliably collect sub-process errors and get cancellation logic right (even when using `multiprocessing`).

So the only motivation now for `open_process()` use directly is a non-SC process-lives-longer then "current nursery" type scenario yes? This `run_process()` call (presuming `.start()` use) is effectively the same but with guaranteed shielded supervision on teardown.

My only last question is **if** the async context mng interface is deprecated then is there now no longer a way to *manually cancel only a specific process*?

Afaict here, you can't force the teardown (634 onwards) without cancelling the task (but that usually requires cancelling a group scope since, nursery, and a new cancel scope isn't created internally here per process). So if cancellation is the new `__aexit__()`, then we somewhat lose granular control of per-process cancellation and/or its completion detection right? Unless of course you want to use `open_process()` directly (but then in that case shouldn't `__aexit__()` code get moved to the post-`yield` body of `open_process()`)?

The streams closing and process killing in [`.aclose()`](https://github.com/python-trio/trio/blob/c0dbe792095b67655cb14deee42d4a050625846b/trio/_subprocess.py#L182) is something we currently depend on in `tractor`.

With that last question, forgive me if I'm missing something.

Reply:
> My only last question is if the async context mng interface is deprecated then is there now no longer a way to manually kill only a specific process?

You can still kill a process. `Process.terminate`/`Process.kill`/`Process.send_signal` are all still available, and you can still close the streams if you want. `Process.aclose` was always just a convenience wrapper around other existing public APIs, and that hasn't changed.

Reply:
> Process.aclose was always just a convenience wrapper around other existing public APIs, and that hasn't changed.

Ok so it will henceforth follow that users *can* roll their own version of `.aclose()` but that won't be default behavior.